finesse/gui/temp_control.py

```diff
@@ -0,0 +1,359 @@
+
+        Args:
+            #x: time
+            #y: temperature
+            # probably won't be inputs since obtainable from elsewhere
+        Returns:
+            None
+        """
+        xdata = list(self._bb_hot_line[0].get_xdata())
+        y1data = list(self._bb_hot_line[0].get_ydata())
+        y2data = list(self._bb_cold_line[0].get_ydata())
+
+        xdata.pop(0)
+        y1data.pop(0)
+        y2data.pop(0)
+
+        x = xdata[-1] + 10
+
+        # Basic RNG for testing
+        y1 = (60 * y1data[-1] + 50) % 70
+        y2 = (5 * y2data[-1] + 1) % 8
+
+        xdata.append(x)
+        y1data.append(y1)
+        y2data.append(y2)
+
+        self._bb_hot_line[0].set_xdata(xdata)
+        self._bb_hot_line[0].set_ydata(y1data)
+        self._bb_cold_line[0].set_xdata(xdata)
+        self._bb_cold_line[0].set_ydata(y2data)
+
+        self._ax1.relim()
+        self._ax2.xaxis.axes.relim()
+        self._ax1.autoscale()
+        self._ax2.xaxis.axes.autoscale()
+        self._ax2.set_ylim([self._ax2.get_ylim()[0], 2 * self._ax2.get_ylim()[1]])
+        self._canvas.draw()
+
+
+class DP9800(QGroupBox):
+    """Widgets to view the DP9800 properties."""
+
+    def __init__(self) -> None:
+        """Creates the widgets to monitor DP9800.
+
+        Returns:
+            None
+        """
+        super().__init__("DP9800")
+
+        layout = self._create_controls()
+        self.setLayout(layout)
+
+    def _create_controls(self) -> QGridLayout:
+
+        layout = QGridLayout()
+
+        layout.addWidget(QLabel("Pt 100"), 1, 0)
+        for i in range(1, 9):
+            wid = QLabel("CH_%d" % i)
+            wid.setAlignment(Qt.AlignmentFlag.AlignCenter)
+            layout.addWidget(wid, 0, i)
+
+        self._ch1 = QLineEdit()
+        self._ch1.setReadOnly(True)
+        layout.addWidget(self._ch1, 1, 1)
+
+        self._ch2 = QLineEdit()
+        self._ch2.setReadOnly(True)
+        layout.addWidget(self._ch2, 1, 2)
+
+        self._ch3 = QLineEdit()
+        self._ch3.setReadOnly(True)
+        layout.addWidget(self._ch3, 1, 3)
+
+        self._ch4 = QLineEdit()
+        self._ch4.setReadOnly(True)
+        layout.addWidget(self._ch4, 1, 4)
+
+        self._ch5 = QLineEdit()
+        self._ch5.setReadOnly(True)
+        layout.addWidget(self._ch5, 1, 5)
+
+        self._ch6 = QLineEdit()
+        self._ch6.setReadOnly(True)
+        layout.addWidget(self._ch6, 1, 6)
+
+        self._ch7 = QLineEdit()
+        self._ch7.setReadOnly(True)
+        layout.addWidget(self._ch7, 1, 7)
+
+        self._ch8 = QLineEdit()
+        self._ch8.setReadOnly(True)
+        layout.addWidget(self._ch8, 1, 8)
+
+        wid = QLabel("POLL")
+        wid.setAlignment(Qt.AlignmentFlag.AlignRight | Qt.AlignmentFlag.AlignVCenter)
+        layout.addWidget(wid, 0, 9, 2, 1)
+
+        self._poll_light = QLabel()
+        self._poll_light.setPixmap(QPixmap("./finesse/gui/images/poll_off.png"))
```

Comment:
You also *might* want to consider creating all these `QPixmaps` in one place and making them e.g. global variables, so you don't have to keep loading the files. It's not important though, so don't worry if it's a faff.

Reply:
Ah yes, I meant to ask about the best way to set the image file path in the PR actually - thanks!

Reply:
I had a go at doing this with `importlib` (which is apparently intended as a replacement for `pkg_resources`) but it was failing the GitHub `pytest` (although not my local `pytest`) so I've removed it for the time being until I can work out how to do it properly. I do therefore still have the hard-coded paths in there, but can try and sort that before merging if you'd prefer.

Reply:
This does need to be sorted before we can merge unfortunately because (I think) it'll lead to broken executables being generated by `pyinstaller`, which we need. (If you want to check whether it works locally, you can run `pyinstaller FINESSE.spec` then run whatever it puts in `dist/`.) I'd do all the other bits first though and then maybe we can tackle this together.

What were the errors that GitHub was reporting? Your penultimate commit on this branch was failing but when I looked into it it seemed to be a spurious test failure on Windows so I reran it and it worked the second time.

Reply:
I think I've finally sorted this out. In case it's useful for future reference, here's what I had to do:

I'm using `importlib`, which seems to be the preferred way to include static files now. One significant difference between `importlib` and `pkg_resources` is that the directory containing the static files needs an `__init__.py` to be properly treated as a module. This allowed me to load in the data from the image files using `importlib.resources.read_binary()`, but the image files still weren't being picked up by `PyInstaller`. I therefore needed to add the images "module" as a hiddenimport, and also tell `PyInstaller` to include the image files in the build, placing them in the `images` module.

Reply:
Good work! Sorry it was such a faff :grimacing:

Reply:
Lemme know when you're ready for re-review